feat: initial documentation for Logos Storage (#166)
status-im/ift-docs: docs/storage/libstorage-tutorial.md

---
title: Build a CLI app with Logos Storage
doc_type: procedure
product: storage
topics: []
steps_layout: sectioned
authors:
owner: logos
doc_version: 1

> slug: libstorage-tutorial

Reviewer: nit: this too seems better to rename

# Getting Started with libstorage

> In this tutorial, we will build a simple [libstorage-based](https://github.com/logos-storage/logos-storage-nim) C application which allows sharing files over the Logos Storage network. Since libstorage is a low-level API and its main use case are library and not  application developers, however, we will use a higher-level C wrapper developed for this tutorial which we refer to as [easylibstorage](https://github.com/logos-storage/easylibstorage).

Reviewer 2: ```suggestion
In this tutorial, we will build a simple [libstorage-based](https://github.com/logos-storage/logos-storage-nim) C application which allows sharing files over the Logos Storage network. Since libstorage is a low-level API, its main use case is to be used as a library and wrapped into a higher-level abstraction. As such, we will use a higher-level C wrapper developed for this tutorial called [easylibstorage](https://github.com/logos-storage/easylibstorage).
```

product: storage
topics: [libstorage, file-sharing, c, easylibstorage]
steps_layout: sectioned
authors:
owner: logos
doc_version: 1
slug: libstorage-share-files
---

# Share files over the Logos Storage network using libstorage

#### Get started building a file uploader and downloader with the easylibstorage C wrapper.

This tutorial shows you how to use libstorage — a low-level C API — to upload and download files over the Logos Storage network. Because libstorage is designed as a building block for higher-level libraries, this tutorial uses [easylibstorage](https://github.com/logos-storage/easylibstorage), a higher-level C wrapper developed to simplify the API.

By the end of this tutorial, you will have two CLI applications: one that uploads a local file to the network and one that downloads it by content ID.

> **Before you start**, ensure the following tools are installed:

Reviewer: let's add `curl` and `unzip` as they are defined in the doc too

- `git`
- CMake 3.14 or later
- A C11-compatible compiler

## What to expect

- You will be able to share any local file over the Logos Storage network from a self-hosted node.
- You will be able to retrieve a shared file from any machine using its SPR and CID strings.
- You will be able to verify file integrity after a network round-trip.

## Step 1: Clone and build easylibstorage locally

Clone the easylibstorage repository and fetch prebuilt libstorage binaries using the bundled script.

> **Note**
>
> Use the bundled fetch script rather than building libstorage from source — it is the simplest way to get working binaries for this tutorial.

1. Clone the repository and enter the project directory:

   ```bash
   git clone https://github.com/logos-storage/easylibstorage.git
   cd easylibstorage
   ```

1. Download the prebuilt libstorage binaries:

   ```bash
   ./scripts/fetch-libstorage.sh
   ```

   Confirm the script completes without errors before continuing.

1. Build easylibstorage:

   ```bash
   cmake -B build -S . -DLOGOS_STORAGE_NIM_ROOT=./libstorage

>    cmake --build build

Reviewer: this fails for me:
```
/work/easylibstorage/easystorage.c:232:12: error: too few arguments to function 'storage_destroy'
      return storage_destroy(node);
/work/easylibstorage/./libstorage/library/libstorage.h:386:9: note: declared here
      int storage_destroy(void *ctx, ...);
```

we need to either:
1. add `git checkout 36b0242` as only this commit seems to work;
2. I think better to cut a semver tag on easylibstorage and synchronize with the next `logos-storage-nim` release, and add reference the tag

to me (2) seems much better

   ```

## Step 2: Create the tutorial directory and source files

> Create a `tutorial` folder inside the cloned repository and add the uploader and downloader source files.

Reviewer: nit: this is a proposal AI has come up with, I think might be good to add

```suggestion
The `examples/` directory in the repo already contains a working uploader and downloader; this tutorial walks through writing them from scratch. If you want to skip ahead and run the shipped versions, configure with - `DCOMPILE_TUTORIAL_EXAMPLES=ON` and rebuild.

Create a `tutorial` folder inside the cloned repository and add the uploader and downloader source files.
```

1. Create the tutorial directory:

   ```bash
   mkdir tutorial
   ```

1. Create `tutorial/uploader.c` with the following content:

   ```c
   /* uploader.c: makes a local file available to the Logos Storage network. */

   #include <stdio.h>
   #include <stdlib.h>
   #include "easystorage.h"

   void panic(const char *msg) {
       fprintf(stderr, "Panic: %s\n", msg);
       exit(1);
   }

   void progress(int total, int complete, int status) {
       printf("\r  %d / %d bytes", complete, total);
       fflush(stdout);
   }

   int main(int argc, char *argv[]) {
       if (argc < 2) {
           printf("Usage: %s <filepath>\n", argv[0]);
           exit(1);
       }

>        node_config cfg = {

Reviewer: in the upstream example it uses `8080`
https://github.com/logos-storage/easylibstorage/blob/36b024244435324226cf652bb8ff924a396144f4/examples/uploader.c#L24

```suggestion
       node_config cfg = {
              .api_port = 8080
```

       if (e_storage_download(node, cid, filepath, progress) != RET_OK) panic("Failed to download file");
       e_storage_stop(node);
       e_storage_destroy(node);
   }
   ```

## Step 3: Update the build configuration and compile

Add the uploader and downloader targets to `CMakeLists.txt` and rebuild.

1. Append the following to the end of `CMakeLists.txt` in the `easylibstorage` root:

   ```cmake
   # Tutorial uploader/downloader
   add_executable(uploader tutorial/uploader.c)
   add_executable(downloader tutorial/downloader.c)

   target_link_libraries(uploader PRIVATE easystorage)
   target_link_libraries(downloader PRIVATE easystorage)
   target_link_libraries(uploader PRIVATE ${LIBSTORAGE_PATH})
   target_link_libraries(downloader PRIVATE ${LIBSTORAGE_PATH})
   ```

1. Rebuild the project:

   ```bash
   cmake -B build -S . -DLOGOS_STORAGE_NIM_ROOT=./libstorage
   cmake --build build
   ```

   After the build completes, the `uploader` and `downloader` executables appear under `./build/`.

## Step 4: Upload and download a file with Logos Storage

Run the uploader and then use its output to download the file in a second terminal.

1. Upload a local file:

   ```bash

>    ./build/uploader ./myfile

Reviewer: nit: to make clear that file needs to be created

```suggestion
    echo "hello world!" > myfile
   ./build/uploader ./myfile
```

   ```

>    The program outputs a `Run:` line containing the SPR and CID strings, then waits for input. Leave this terminal open.

Reviewer: ```suggestion
   The program outputs a `Run:` line containing the SPR and CID strings, then waits for input. Leave this terminal open. When this line appears, the upload is complete and the file is being served. Leave this terminal open; the node must keep running for the downloader to fetch the file.
```

> 1. Open a second terminal and paste the `Run:` command from the uploader output. It will look similar to:

Reviewer: ```suggestion
1. Open a second terminal and paste the `Run:` command from the uploader output. If you are working inside a container or remote shell, open a second docker exec/ssh/tmux session into the same environment. It will look similar to:
```

   ```bash
   ./build/downloader spr:<spr_string> <cid_string> ./output-file
   ```

>    Wait for the command to finish. The file is saved to `./output-file`.

Reviewer: ```suggestion
   Wait for the command to finish. The file is saved to `./output-file`.
   
   The downloader exits silently on success; if you see a Panic: line, the download failed. Use `cmp` (step below) as the authoritative success check.
```
